crates/ra_assists/src/auto_import.rs

```diff
@@ -0,0 +1,780 @@
+            Some(ast::PathSegmentKind::CrateKw) => buf.push_str("crate"),
+            None => {}
+        }
+        first = false;
+    }
+}
+
+#[derive(Copy, Clone)]
+enum PathSegmentsMatch {
+    // Patch matches exactly
+    Full,
+    // None of the segments matched. It's a more explicit Partial(0)
+    Empty,
+    // When some of the segments matched
+    Partial(usize),
+    // When all the segments of the right path are matched against the left path,
+    // but the left path is longer.
+    PartialLeft(usize),
+    // When all the segments of the left path are matched against the right path,
+    // but the right path is longer.
+    PartialRight(usize),
+    // In all the three cases above we keep track of how many segments matched
+}
+
+fn compare_path_segments(
+    left: &[&ast::PathSegment],
+    right: &[&ast::PathSegment],
+) -> PathSegmentsMatch {
+    let mut matching = 0;
+    for either_or_both in left.iter().zip_longest(right.iter()) {
+        match either_or_both {
+            EitherOrBoth::Both(left, right) => {
+                if compare_path_segment(left, right) {
+                    matching += 1
+                } else {
+                    return if matching == 0 {
+                        PathSegmentsMatch::Empty
+                    } else {
+                        PathSegmentsMatch::Partial(matching)
+                    };
+                }
+            }
+            EitherOrBoth::Left(_) => {
+                return PathSegmentsMatch::PartialLeft(matching);
+            }
+            EitherOrBoth::Right(_) => {
+                return PathSegmentsMatch::PartialRight(matching);
+            }
+        }
+    }
+    return PathSegmentsMatch::Full;
+}
+
+fn compare_path_segment(a: &ast::PathSegment, b: &ast::PathSegment) -> bool {
+    if let (Some(ka), Some(kb)) = (a.kind(), b.kind()) {
+        return match (ka, kb) {
+            (ast::PathSegmentKind::Name(nameref_a), ast::PathSegmentKind::Name(nameref_b)) => {
+                nameref_a.text() == nameref_b.text()
+            }
+            (ast::PathSegmentKind::SelfKw, ast::PathSegmentKind::SelfKw) => true,
+            (ast::PathSegmentKind::SuperKw, ast::PathSegmentKind::SuperKw) => true,
+            (ast::PathSegmentKind::CrateKw, ast::PathSegmentKind::CrateKw) => true,
+            (_, _) => false,
+        };
+    } else {
+        false
+    }
+}
+
+fn compare_path_segment_with_name(a: &ast::PathSegment, b: &ast::Name) -> bool {
+    if let Some(ka) = a.kind() {
+        return match (ka, b) {
+            (ast::PathSegmentKind::Name(nameref_a), _) => nameref_a.text() == b.text(),
+            (_, _) => false,
+        };
+    } else {
+        false
+    }
+}
+
+#[derive(Copy, Clone)]
+enum ImportAction<'a> {
+    Nothing,
+    // Add a brand new use statement.
+    AddNewUse(
+        Option<&'a SyntaxNode>, // anchor node
+        bool,                   // true if we want to add the new statement after the anchor
```

Comment:
Perhaps it makes sense to make this a struct variant, if we need a comment anyway? Positional bools are awkward .